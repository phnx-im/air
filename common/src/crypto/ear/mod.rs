// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! This module and its submodules contain structs and types to facilitate (EAR)
//! encryption of other structs on the backend. See the individual submodules
//! for details.

pub mod keys;
mod trait_impls;
mod traits;

use std::marker::PhantomData;

use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};
pub use traits::{EarDecryptable, EarEncryptable, EarKey};

use aes_gcm::Aes256Gcm;
use serde::{Deserialize, Serialize};

/// This type determines the AEAD scheme used for encryption at rest (EAR) by
/// the backend.
/// TODO: Replace with a key-committing scheme.
pub type Aead = Aes256Gcm;
/// Key size of the [`Aead`] scheme
pub const AEAD_KEY_SIZE: usize = 32;
const AEAD_NONCE_SIZE: usize = 12;

#[derive(
    Clone,
    Debug,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    sqlx::Type,
)]
#[sqlx(type_name = "aead_ciphertext")]
pub struct AeadCiphertext {
    ciphertext: Vec<u8>,
    nonce: [u8; AEAD_NONCE_SIZE],
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Ciphertext<CT> {
    ct: AeadCiphertext,
    pd: PhantomData<CT>,
}

impl<CT> From<AeadCiphertext> for Ciphertext<CT> {
    fn from(aead_ciphertext: AeadCiphertext) -> Self {
        Self {
            ct: aead_ciphertext,
            pd: PhantomData,
        }
    }
}

impl<CT> From<Ciphertext<CT>> for AeadCiphertext {
    fn from(ciphertext: Ciphertext<CT>) -> Self {
        ciphertext.ct
    }
}

impl AeadCiphertext {
    pub fn new(ciphertext: Vec<u8>, nonce: [u8; AEAD_NONCE_SIZE]) -> Self {
        Self { ciphertext, nonce }
    }

    pub fn into_parts(self) -> (Vec<u8>, [u8; AEAD_NONCE_SIZE]) {
        let Self { ciphertext, nonce } = self;
        (ciphertext, nonce)
    }
}

#[cfg(any(feature = "test_utils", test))]
impl<CT> Ciphertext<CT> {
    pub fn dummy() -> Self {
        AeadCiphertext::dummy().into()
    }

    pub fn random() -> Self {
        AeadCiphertext::random().into()
    }

    pub fn flip_bit(&mut self) {
        self.ct.flip_bit();
    }
}

#[cfg(any(feature = "test_utils", test))]
impl AeadCiphertext {
    pub fn dummy() -> Self {
        Self {
            ciphertext: vec![1u8; 32],
            nonce: [1u8; AEAD_NONCE_SIZE],
        }
    }

    pub fn random() -> Self {
        use rand::Rng;

        let mut rng = rand::thread_rng();
        Self {
            ciphertext: rng.r#gen::<[u8; 32]>().into(),
            nonce: rng.r#gen::<[u8; AEAD_NONCE_SIZE]>(),
        }
    }

    fn flip_bit(&mut self) {
        let byte = self.ciphertext.pop().unwrap();
        self.ciphertext.push(byte ^ 1);
    }
}
