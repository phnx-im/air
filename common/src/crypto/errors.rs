// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use super::*;

#[derive(Error, Debug, PartialEq, Eq, Clone)]
pub enum KeyGenerationError {
    /// Error generating signature keypair
    #[error("Error generating signature keypair")]
    KeypairGeneration,
}

#[derive(Debug, Error)]
pub enum RandomnessError {
    #[error("Insufficient randomness")]
    InsufficientRandomness,
}

#[derive(Error, Debug, Clone)]
pub enum DecryptionError {
    /// Error decrypting ciphertext.
    #[error("Error decrypting ciphertext.")]
    DecryptionError,
    /// Error deserializing payload.
    #[error("Error deserializing payload.")]
    DeserializationError,
    /// Error serializing AAD.
    #[error("Error serializing AAD.")]
    SerializationError,
    /// Error ratcheting forward
    #[error("Error ratcheting forward")]
    RatchetForwardError(#[from] LibraryError),
}

#[derive(Error, Debug, Clone)]
pub enum EncryptionError {
    /// Not enough randomness to generate Nonce
    #[error("Not enough randomness to generate Nonce")]
    RandomnessError,
    /// Encryption error
    #[error("Encryption error")]
    EncryptionError,
    /// Codec error
    #[error("Codec error")]
    SerializationError,
    /// Error ratcheting forward
    #[error("Error ratcheting forward")]
    RatchetForwardError(#[from] LibraryError),
}
