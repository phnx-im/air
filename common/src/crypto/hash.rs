// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::marker::PhantomData;

use sha2::{Digest, Sha256};
use tls_codec::{Serialize, TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::{crypto::Labeled, identifiers::TlsStr};

pub type HashAlg = Sha256;
pub const HASH_SIZE: usize = 32;

#[derive(Debug, TlsSize, TlsSerialize)]
struct LabeledHashPayload<'a, T: Serialize> {
    operation: TlsStr<'a>,
    label: TlsStr<'a>,
    payload: T,
}

impl<T: Serialize + Labeled> LabeledHashPayload<'_, T> {
    fn new(payload: T) -> Self {
        Self {
            operation: TlsStr("Infra Hash"),
            label: TlsStr(T::LABEL),
            payload,
        }
    }
}

#[derive(TlsDeserializeBytes, TlsSerialize, TlsSize, serde::Serialize, serde::Deserialize)]
#[serde(transparent)]
pub struct Hash<T: Labeled + Sized> {
    bytes: [u8; HASH_SIZE],
    _marker: PhantomData<T>,
}

impl<T: Labeled + Sized> Hash<T> {
    #[cfg(any(feature = "test_utils", test))]
    pub fn new_for_test(mut bytes: Vec<u8>) -> Self {
        // Pad until the length is exactly HASH_SIZE
        if bytes.len() < HASH_SIZE {
            bytes.resize(HASH_SIZE, 0);
        }
        // Ensure the length is exactly HASH_SIZE
        assert_eq!(
            bytes.len(),
            HASH_SIZE,
            "Hash must be exactly {} bytes long",
            HASH_SIZE
        );
        let mut hash_bytes = [0u8; HASH_SIZE];
        hash_bytes.copy_from_slice(&bytes[..HASH_SIZE]);
        Self {
            bytes: hash_bytes,
            _marker: PhantomData,
        }
    }

    pub fn as_bytes(&self) -> &[u8; HASH_SIZE] {
        &self.bytes
    }

    pub fn into_bytes(&self) -> [u8; HASH_SIZE] {
        self.bytes
    }

    pub fn from_bytes(bytes: [u8; HASH_SIZE]) -> Self {
        Self {
            bytes,
            _marker: PhantomData,
        }
    }
}

/// Trait for types that can be hashed
///
/// Default implementation requires the type (and references to it) to implement
/// `tls_codec::Serialize` and `Labeled`.
#[allow(private_bounds)]
pub trait Hashable: HashableHelper {
    /// Returns the hash of the data as a byte array.
    fn hash(&self) -> Hash<Self> {
        HashableHelper::hash(self)
    }
}

trait HashableHelper: Labeled + Sized {
    /// Returns the hash of the data as a byte array.
    fn hash(&self) -> Hash<Self>;
}

impl<T: Labeled + Sized> HashableHelper for T
where
    for<'a> &'a T: Serialize,
{
    fn hash(&self) -> Hash<T> {
        let serialization_result = LabeledHashPayload::new(self).tls_serialize_detached();
        debug_assert!(serialization_result.is_ok(), "Serialization failed");
        let serialized_labeled_payload = serialization_result.unwrap_or_default();

        let hash_bytes = HashAlg::digest(&serialized_labeled_payload);
        Hash {
            bytes: hash_bytes.into(),
            _marker: PhantomData,
        }
    }
}

mod trait_impls {
    use sqlx::{Database, Encode, Type};

    use super::*;

    impl<T: Labeled> std::fmt::Debug for Hash<T> {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            write!(f, "Hash<{}>({:?})", T::LABEL, self.bytes)
        }
    }

    impl<T: Labeled> PartialEq for Hash<T> {
        fn eq(&self, other: &Self) -> bool {
            self.bytes == other.bytes
        }
    }

    impl<T: Labeled> Eq for Hash<T> {}

    impl<T: Labeled> std::hash::Hash for Hash<T> {
        fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
            self.bytes.hash(state);
        }
    }

    impl<T: Labeled> Clone for Hash<T> {
        fn clone(&self) -> Self {
            Self {
                bytes: self.bytes.clone(),
                _marker: PhantomData,
            }
        }
    }

    impl<T: Labeled> Copy for Hash<T> {}

    impl<T: Labeled, DB: Database> Type<DB> for Hash<T>
    where
        Vec<u8>: Type<DB>,
    {
        fn type_info() -> <DB as Database>::TypeInfo {
            <Vec<u8> as Type<DB>>::type_info()
        }
    }

    impl<'a, T: Labeled, DB: Database> Encode<'a, DB> for Hash<T>
    where
        Vec<u8>: Encode<'a, DB>,
    {
        fn encode_by_ref(
            &self,
            buf: &mut <DB as Database>::ArgumentBuffer<'a>,
        ) -> Result<sqlx::encode::IsNull, sqlx::error::BoxDynError> {
            let bytes = self.as_bytes().to_vec();
            Encode::<DB>::encode(&bytes, buf)
        }
    }

    impl<'a, T: Labeled, DB: Database> sqlx::Decode<'a, DB> for Hash<T>
    where
        Vec<u8>: sqlx::Decode<'a, DB>,
    {
        fn decode(value: <DB as Database>::ValueRef<'a>) -> Result<Self, sqlx::error::BoxDynError> {
            let bytes_vec = <Vec<u8>>::decode(value)?;
            let bytes: [u8; HASH_SIZE] = bytes_vec
                .try_into()
                .map_err(|_| sqlx::error::BoxDynError::from("Invalid hash length"))?;
            Ok(Self {
                bytes,
                _marker: PhantomData,
            })
        }
    }
}
