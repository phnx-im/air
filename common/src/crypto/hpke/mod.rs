// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    openmls::{
        key_packages::InitKey,
        prelude::{HpkeAeadType, HpkeConfig, HpkeKdfType, HpkeKemType, HpkePublicKey},
    },
    openmls_rust_crypto::OpenMlsRustCrypto,
    openmls_traits::{
        OpenMlsProvider,
        crypto::OpenMlsCrypto,
        random::OpenMlsRand,
        types::{HpkeCiphertext, HpkePrivateKey},
    },
};
use serde::{Deserialize, Serialize};
use tls_codec::Serialize as TlsSerializeTrait;
use tracing::error;

use crate::identifiers::{ClientConfig, SealedClientReference};

use super::{
    RawKey,
    errors::{DecryptionError, EncryptionError, RandomnessError},
    secrets::SecretBytes,
};

pub mod trait_impls;

#[derive(Debug, Serialize, Deserialize)]
#[serde(transparent)]
pub struct EncryptionKey<KT> {
    key: Vec<u8>,
    _type: std::marker::PhantomData<KT>,
}

pub const HPKE_CONFIG: HpkeConfig = HpkeConfig(
    HpkeKemType::DhKem25519,
    HpkeKdfType::HkdfSha256,
    HpkeAeadType::AesGcm256,
);

impl<KT> EncryptionKey<KT> {
    fn new(bytes: Vec<u8>) -> Self {
        Self {
            key: bytes,
            _type: std::marker::PhantomData,
        }
    }

    /// Encrypt the given plaintext using this key.
    pub(crate) fn encrypt(&self, info: &[u8], aad: &[u8], plain_txt: &[u8]) -> HpkeCiphertext {
        let rust_crypto = OpenMlsRustCrypto::default();
        rust_crypto
            .crypto()
            .hpke_seal(HPKE_CONFIG, &self.key, info, aad, plain_txt)
            // TODO: get rid of unwrap
            .unwrap()
    }

    #[cfg(any(test, feature = "test_utils"))]
    pub fn new_for_test(bytes: Vec<u8>) -> Self {
        Self::new(bytes)
    }
}

impl<KT: RawKey> EncryptionKey<KT> {
    pub fn into_bytes(self) -> Vec<u8> {
        self.key
    }

    pub fn from_bytes(bytes: Vec<u8>) -> Self {
        Self {
            key: bytes,
            _type: std::marker::PhantomData,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DecryptionKey<KT> {
    decryption_key: SecretBytes,
    #[serde(bound = "")]
    encryption_key: EncryptionKey<KT>,
}

#[cfg(any(test, feature = "test_utils"))]
impl<KT> PartialEq for DecryptionKey<KT> {
    fn eq(&self, other: &Self) -> bool {
        let decryption_key: &[u8] = self.decryption_key.as_ref();
        let other_decryption_key: &[u8] = other.decryption_key.as_ref();
        decryption_key == other_decryption_key && self.encryption_key == other.encryption_key
    }
}

#[cfg(any(test, feature = "test_utils"))]
impl<KT> Eq for DecryptionKey<KT> {}

impl<KT> DecryptionKey<KT> {
    pub fn new(decryption_key: HpkePrivateKey, encryption_key: EncryptionKey<KT>) -> Self {
        Self {
            decryption_key: decryption_key.as_ref().to_vec().into(),
            encryption_key,
        }
    }

    pub fn decrypt(
        &self,
        info: &[u8],
        aad: &[u8],
        ct: &HpkeCiphertext,
    ) -> Result<Vec<u8>, DecryptionError> {
        let rust_crypto = OpenMlsRustCrypto::default();
        rust_crypto
            .crypto()
            .hpke_open(HPKE_CONFIG, ct, &self.decryption_key, info, aad)
            .map_err(|_| DecryptionError::DecryptionError)
    }

    pub fn generate() -> Result<Self, RandomnessError> {
        let provider = OpenMlsRustCrypto::default();
        let key_seed = provider
            .rand()
            .random_array::<32>()
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        let keypair = provider
            .crypto()
            .derive_hpke_keypair(HPKE_CONFIG, &key_seed)
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        Ok(Self::new(
            keypair.private,
            EncryptionKey::new(keypair.public),
        ))
    }

    pub fn encryption_key(&self) -> &EncryptionKey<KT> {
        &self.encryption_key
    }
}

// TODO: We might want to properly fix AAD and Info at some point.
pub trait HpkeEncryptable<KT, HpkeCiphertextType: From<HpkeCiphertext>>:
    tls_codec::Serialize
{
    /// Encrypts the data with the given encryption key.
    fn encrypt(
        &self,
        encryption_key: &EncryptionKey<KT>,
        info: &[u8],
        aad: &[u8],
    ) -> HpkeCiphertextType {
        // Hiding a LibraryError behind an empty vec.
        let plain_txt = self.tls_serialize_detached().unwrap_or_default();
        encryption_key.encrypt(info, aad, &plain_txt).into()
    }
}

pub trait HpkeDecryptable<KT, HpkeCiphertextType: AsRef<HpkeCiphertext>>:
    tls_codec::DeserializeBytes + Sized
{
    fn decrypt(
        ct: HpkeCiphertextType,
        decryption_key: &DecryptionKey<KT>,
        info: &[u8],
        aad: &[u8],
    ) -> Result<Self, DecryptionError> {
        let plaintext = decryption_key.decrypt(info, aad, ct.as_ref())?;
        Self::tls_deserialize_exact_bytes(&plaintext).map_err(|e| {
            error!(%e, "Error deserializing decrypted data");
            DecryptionError::DeserializationError
        })
    }
}

pub struct JoinerInfoKeyType;
pub type JoinerInfoEncryptionKey = EncryptionKey<JoinerInfoKeyType>;

// We need this From trait, because we have to work with the hpke init key from
// the KeyPackage, which we get as HpkePublicKey from OpenMLS.
impl From<HpkePublicKey> for JoinerInfoEncryptionKey {
    fn from(value: HpkePublicKey) -> Self {
        Self::new(value.as_slice().to_vec())
    }
}

impl From<InitKey> for JoinerInfoEncryptionKey {
    fn from(value: InitKey) -> Self {
        Self::new(value.as_slice().to_vec())
    }
}

pub type JoinerInfoDecryptionKey = DecryptionKey<JoinerInfoKeyType>;

// We need this From trait, because we have to work with the hpke init key from
// the KeyPackage, which we get as HpkePrivateKey and HpkePublicKey from
// OpenMLS.
impl From<(HpkePrivateKey, InitKey)> for JoinerInfoDecryptionKey {
    fn from((sk, init_key): (HpkePrivateKey, InitKey)) -> Self {
        let vec: Vec<u8> = init_key.key().as_slice().to_vec();
        DecryptionKey::new(sk, EncryptionKey::new(vec))
    }
}

#[derive(Debug)]
pub struct ClientIdKeyType;
pub type ClientIdEncryptionKey = EncryptionKey<ClientIdKeyType>;

impl RawKey for ClientIdKeyType {}

impl ClientIdEncryptionKey {
    pub fn seal_client_config(
        &self,
        client_config: ClientConfig,
    ) -> Result<SealedClientReference, EncryptionError> {
        let bytes = client_config
            .tls_serialize_detached()
            .map_err(|_| EncryptionError::SerializationError)?;
        let ciphertext = self.encrypt(&[], &[], &bytes);
        Ok(SealedClientReference { ciphertext })
    }
}

pub type ClientIdDecryptionKey = DecryptionKey<ClientIdKeyType>;
