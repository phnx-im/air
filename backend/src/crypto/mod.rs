// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! A module that provides various traits and structs that allow the use of
//! cryptographic primitives such as AEAD, MACs and KDFs.
//!
//! TODO: Once const-generics allows the use of enums, we could get rid of a
//! number of structs and boilerplate code.
//! TODO: A proper RNG provider for use with all crypto functions that require
//! randomness, i.e. mainly secret and nonce sampling.
#![allow(unused_variables)]
use argon2::Argon2;
use chrono::{DateTime, Utc};
use hpke::{Hpke, HpkePrivateKey, HpkePublicKey};
use hpke_rs_crypto::types::{AeadAlgorithm, KdfAlgorithm, KemAlgorithm};
use hpke_rs_rust_crypto::HpkeRustCrypto;
use opaque_ke::CipherSuite;
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use tls_codec::{TlsDeserialize, TlsSerialize, TlsSize};
use utoipa::ToSchema;

use crate::{
    crypto::ear::EarEncryptable,
    messages::{client_ds::QueueMessagePayload, QueueMessage},
    qs::SealedClientReference,
    LibraryError,
};

use self::{
    ear::{keys::RatchetKey, Ciphertext, EncryptionError},
    kdf::{keys::RatchetSecret, KdfDerivable},
};

/// Default ciphersuite we use for OPAQUE
pub struct OpaqueCiphersuite;

impl CipherSuite for OpaqueCiphersuite {
    type OprfCs = opaque_ke::Ristretto255;
    type KeGroup = opaque_ke::Ristretto255;
    type KeyExchange = opaque_ke::key_exchange::tripledh::TripleDh;

    type Ksf = Argon2<'static>;
}

// The OPAQUE ciphersuite's Noe: The size of a serialized OPRF group element output from SerializeElement.
const OPAQUE_NOE: usize = 32;
// The OPAQUE ciphersuite's Nok: The size of an OPRF private key as output from DeriveKeyPair.
const OPAQUE_NOK: usize = 32;
// The OPAQUE ciphersuite's Nn: Nonce length.
const OPAQUE_NN: usize = 32;
// The OPAQUE ciphersuite's Nm: MAC length.
const OPAQUE_NM: usize = 32;
// The OPAQUE ciphersuite's Nh: Hash length.
const OPAQUE_NH: usize = 32;
// The OPAQUE ciphersuite's Npk: Public key length.
const OPAQUE_NPK: usize = 32;

// The size of an OPAQUE envelope (Nn + nM)
const OPAQUE_ENVELOPE_SIZE: usize = OPAQUE_NN + OPAQUE_NM;
const OPAQUE_CREDENTIAL_REQUEST_SIZE: usize = OPAQUE_NOE;
const OPAQUE_CREDENTIAL_RESPONSE_SIZE: usize =
    OPAQUE_NOE + OPAQUE_NN + OPAQUE_NPK + OPAQUE_NN + OPAQUE_NM;
const OPAQUE_AUTH_REQUEST_SIZE: usize = OPAQUE_NN + OPAQUE_NPK;
const OPAQUE_AUTH_RESPONSE_SIZE: usize = OPAQUE_NN + OPAQUE_NPK + OPAQUE_NM;

// The size of the blinded message, i.e. a serialized OPRF group element using the
// ciphersuite defined above.
pub(crate) const OPAQUE_REGISTRATION_REQUEST_SIZE: usize = OPAQUE_NOE;
// The size of the evaluated message, i.e. a serialized OPRF group element, plus that of the server public key using the
// ciphersuite defined above.
pub(crate) const OPAQUE_REGISTRATION_RESPONSE_SIZE: usize = OPAQUE_NOE + OPAQUE_NOK;
// The size of the client upload after successful registration: The client public key, as well as a masking key and an envelope.
pub(crate) const OPAQUE_REGISTRATION_RECORD_SIZE: usize =
    OPAQUE_NPK + OPAQUE_NH + OPAQUE_ENVELOPE_SIZE;

// The size of the KE1 struct
pub(crate) const OPAQUE_LOGIN_REQUEST_SIZE: usize =
    OPAQUE_CREDENTIAL_REQUEST_SIZE + OPAQUE_AUTH_REQUEST_SIZE;
// The size of the KE2 struct
pub(crate) const OPAQUE_LOGIN_RESPONSE_SIZE: usize =
    OPAQUE_CREDENTIAL_RESPONSE_SIZE + OPAQUE_AUTH_RESPONSE_SIZE;
// The size of the KE3 struct
pub(crate) const OPAQUE_LOGIN_FINISH_SIZE: usize = OPAQUE_NM;

/// This type determines the hash function used by the backend.
pub type Hash = Sha256;

pub mod ear;
pub mod kdf;
pub mod mac;
pub mod secrets;
pub(super) mod serde_arrays;
pub mod signatures;

#[derive(Debug)]
pub enum RandomnessError {
    InsufficientRandomness,
}

#[derive(Clone)]
pub struct EncryptedDsGroupState {
    pub ciphertext: Ciphertext,
    pub last_used: DateTime<Utc>,
    pub deleted_queues: Vec<SealedClientReference>,
}

impl From<Ciphertext> for EncryptedDsGroupState {
    fn from(ciphertext: Ciphertext) -> Self {
        Self {
            ciphertext,
            last_used: Utc::now(),
            deleted_queues: Vec::new(),
        }
    }
}

impl AsRef<Ciphertext> for EncryptedDsGroupState {
    fn as_ref(&self) -> &Ciphertext {
        &self.ciphertext
    }
}

pub(crate) type RatchetKeyUpdate = Vec<u8>;

#[derive(Serialize, Deserialize, Clone, Debug, TlsSerialize, TlsDeserialize, TlsSize)]
pub struct QueueRatchet {
    sequence_number: u64,
    secret: RatchetSecret,
    key: RatchetKey,
}

// TODO: Implement the ratchet key.
impl QueueRatchet {
    /// Initialize a new ratchet key.
    pub fn random() -> Result<Self, RandomnessError> {
        let secret = RatchetSecret::random()?;
        let key = RatchetKey::derive(&secret, Vec::new())
            .map_err(|_| RandomnessError::InsufficientRandomness)?;
        Ok(Self {
            sequence_number: 0,
            secret,
            key,
        })
    }

    /// Encrypt the given payload.
    pub fn encrypt(
        &mut self,
        payload: QueueMessagePayload,
    ) -> Result<QueueMessage, EncryptionError> {
        // TODO: We want domain separation: FQDN, UserID & ClientID.
        let ciphertext = payload.encrypt(&self.key)?;

        let secret = RatchetSecret::derive(&self.secret, Vec::new())
            .map_err(|_| EncryptionError::LibraryError)?;
        let key =
            RatchetKey::derive(&secret, Vec::new()).map_err(|_| EncryptionError::LibraryError)?;

        self.secret = secret;
        self.key = key;
        self.sequence_number += 1;

        Ok(QueueMessage {
            sequence_number: self.sequence_number,
            ciphertext,
        })
    }

    /// Decrypt the given payload.
    pub fn decrypt(&self, queue_message: QueueMessage) -> Vec<u8> {
        todo!()
    }

    /// Sample some fresh entropy and inject it into the current key. Returns the entropy.
    pub fn update(&mut self) -> RatchetKeyUpdate {
        todo!()
    }

    /// Get the current sequence number
    pub fn sequence_number(&self) -> u64 {
        self.sequence_number
    }
}

#[derive(Serialize, Deserialize, ToSchema, Clone, TlsSerialize, TlsDeserialize, TlsSize)]
pub struct HpkeCiphertext {
    pub kem_output: Vec<u8>,
    pub ciphertext: Vec<u8>,
}

#[derive(Clone, Serialize, Deserialize, ToSchema, Debug, TlsSerialize, TlsDeserialize, TlsSize)]
pub struct EncryptionPublicKey {
    public_key: HpkePublicKey,
}

impl From<Vec<u8>> for EncryptionPublicKey {
    fn from(value: Vec<u8>) -> Self {
        Self {
            public_key: value.into(),
        }
    }
}

impl EncryptionPublicKey {
    /// Encrypt the given plaintext using this key.
    pub(crate) fn encrypt(
        &self,
        info: &[u8],
        aad: &[u8],
        plain_txt: &[u8],
    ) -> Result<HpkeCiphertext, LibraryError> {
        Hpke::<HpkeRustCrypto>::new(
            hpke::Mode::Base,
            KemAlgorithm::DhKem25519,
            KdfAlgorithm::HkdfSha256,
            AeadAlgorithm::Aes256Gcm,
        )
        .seal(&self.public_key, info, aad, plain_txt, None, None, None)
        .map_err(|_| LibraryError::unexpected_crypto_error("Error encrypting with HPKE."))
        .map(|(kem_output, ciphertext)| HpkeCiphertext {
            kem_output,
            ciphertext,
        })
    }
}

pub enum DecryptionError {
    DecryptionError,
}

#[derive(Debug)]
pub struct DecryptionPrivateKey {
    private_key: HpkePrivateKey,
}

impl DecryptionPrivateKey {
    pub(crate) fn decrypt(
        &self,
        info: &[u8],
        aad: &[u8],
        ct: &HpkeCiphertext,
    ) -> Result<Vec<u8>, DecryptionError> {
        Hpke::<HpkeRustCrypto>::new(
            hpke::Mode::Base,
            KemAlgorithm::DhKem25519,
            KdfAlgorithm::HkdfSha256,
            AeadAlgorithm::Aes256Gcm,
        )
        .open(
            &ct.kem_output,
            &self.private_key,
            info,
            aad,
            &ct.ciphertext,
            None,
            None,
            None,
        )
        .map_err(|_| DecryptionError::DecryptionError)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema, TlsSerialize, TlsDeserialize, TlsSize)]
pub struct RatchetPublicKey {
    encryption_key: EncryptionPublicKey,
}
