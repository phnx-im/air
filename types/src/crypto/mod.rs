// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! A module that provides various traits and structs that allow the use of
//! cryptographic primitives such as AEAD, MACs and KDFs.
//!
//! TODO: Once const-generics allows the use of enums, we could get rid of a
//! number of structs and boilerplate code.
//! TODO: A proper RNG provider for use with all crypto functions that require
//! randomness, i.e. mainly secret and nonce sampling.
#![allow(unused_variables)]
use std::marker::PhantomData;

use hpke::{DecryptionKey, EncryptionKey};
use kdf::keys::ConnectionKeyType;
use serde::{Deserialize, Serialize};
use sha2::Sha256;
use thiserror::Error;
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::{LibraryError, crypto::ear::EarEncryptable, messages::QueueMessage};

use self::{
    ear::{EarDecryptable, keys::RatchetKey},
    errors::RandomnessError,
    kdf::{KdfDerivable, keys::RatchetSecret},
};

pub use opaque::OpaqueCiphersuite;

/// This type determines the hash function used by the backend.
pub type Hash = Sha256;

pub mod ear;
pub mod errors;
pub mod hpke;
pub mod indexed_aead;
pub mod kdf;
pub mod opaque;
pub mod ratchet;
pub mod secrets;
pub(super) mod serde_arrays;
pub mod signatures;

/// Marker trait for keys that can be converted to and from raw bytes
pub trait RawKey {}

pub type RatchetKeyUpdate = Vec<u8>;

#[derive(Debug)]
pub struct RatchetKeyType;
pub type RatchetEncryptionKey = EncryptionKey<RatchetKeyType>;

impl RawKey for RatchetKeyType {}

pub type RatchetDecryptionKey = DecryptionKey<RatchetKeyType>;

pub type ConnectionEncryptionKey = EncryptionKey<ConnectionKeyType>;
pub type ConnectionDecryptionKey = DecryptionKey<ConnectionKeyType>;

#[cfg(test)]
mod test {
    use crate::codec::PhnxCodec;

    use super::*;

    #[test]
    fn encryption_key_serde_codec() {
        let key = RatchetEncryptionKey::new_for_test(vec![1, 2, 3]);
        insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&key).unwrap());
    }

    #[test]
    fn encryption_key_serde_json() {
        let key = RatchetEncryptionKey::new_for_test(vec![1, 2, 3]);
        insta::assert_json_snapshot!(key);
    }
}
