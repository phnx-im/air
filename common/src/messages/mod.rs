// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    openmls::prelude::OpenMlsRand, openmls_rust_crypto::OpenMlsRustCrypto,
    openmls_traits::OpenMlsProvider,
};
use serde::{Deserialize, Serialize};
use tls_codec::{TlsDeserializeBytes, TlsSerialize, TlsSize};

use crate::crypto::{
    ear::{AeadCiphertext, Ciphertext},
    errors::RandomnessError,
};

pub mod client_as;
pub mod client_as_out;
pub mod client_ds;
pub mod client_ds_out;
pub mod client_qs;
pub mod connection_package;
pub mod push_token;
pub mod welcome_attribution_info;

#[derive(
    Serialize,
    Deserialize,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    PartialEq,
    Eq,
    Clone,
    Debug,
    sqlx::Type,
)]
#[sqlx(transparent)]
pub struct FriendshipToken(Vec<u8>);

impl FriendshipToken {
    pub fn from_bytes(token: Vec<u8>) -> Self {
        Self(token)
    }

    pub fn into_bytes(self) -> Vec<u8> {
        self.0
    }

    pub fn random() -> Result<Self, RandomnessError> {
        let token = OpenMlsRustCrypto::default()
            .rand()
            .random_vec(32)
            .map_err(|_| RandomnessError::InsufficientRandomness)?;

        Ok(Self(token))
    }

    pub fn token(&self) -> &[u8] {
        self.0.as_ref()
    }
}

/// Enum encoding the version of the MlsInfra protocol that was used to create
/// the given message.
#[derive(
    Debug,
    TlsSerialize,
    TlsDeserializeBytes,
    TlsSize,
    Clone,
    Copy,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
)]
#[repr(u8)]
pub enum MlsInfraVersion {
    Alpha,
}

impl Default for MlsInfraVersion {
    fn default() -> Self {
        Self::Alpha
    }
}

// === Queue ===

#[derive(
    Clone, Debug, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserializeBytes, TlsSize,
)]
pub struct QueueMessage {
    pub sequence_number: u64,
    pub ciphertext: AeadCiphertext,
}

#[derive(Debug, Clone)]
pub struct EncryptedQsQueueMessageCtype;

pub type EncryptedQsQueueMessage = Ciphertext<EncryptedQsQueueMessageCtype>;

#[derive(Debug, Clone)]
pub struct EncryptedAsQueueMessageCtype;

pub type EncryptedAsQueueMessage = Ciphertext<EncryptedAsQueueMessageCtype>;

#[cfg(test)]
mod test {
    use crate::{codec::AirCodec, crypto::ear::AeadCiphertext};

    use super::*;

    #[test]
    fn test_queue_message_serde_codec() {
        let message = QueueMessage {
            sequence_number: 1,
            ciphertext: AeadCiphertext::dummy(),
        };
        insta::assert_binary_snapshot!(".cbor", AirCodec::to_vec(&message).unwrap());
    }

    #[test]
    fn test_queue_message_serde_json() {
        let message = QueueMessage {
            sequence_number: 1,
            ciphertext: AeadCiphertext::dummy(),
        };
        insta::assert_json_snapshot!(message);
    }
}
