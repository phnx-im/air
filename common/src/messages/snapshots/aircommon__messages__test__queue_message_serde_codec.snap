---
source: common/src/messages/mod.rs
expression: "PersistenceCodec::to_vec(&message).unwrap()"
extension: cbor
snapshot_kind: binary
---
